dlib/image_transforms/draw.h

```diff
@@ -534,44 +533,95 @@ namespace dlib
                     rgb_alpha_pixel temp = alpha_pixel;
                     temp.alpha = max_alpha-static_cast<unsigned char>((p.x()-right_boundary[i])*max_alpha);
                     if (valid_area.contains(p))
-                        assign_pixel(image(p.y(), p.x()),temp);
+                        assign_pixel(img[p.y()][p.x()],temp);
                 }
             }
             else if (delta < 0) // on the top side
             {
                 for (long x = static_cast<long>(std::floor(right_boundary[i-1]))+1; x <= right_x; ++x)
                 {
-                    const point p(x, i+top-1);
+                    const point p(x, i + valid_area.top() - 1);
                     rgb_alpha_pixel temp = alpha_pixel;
                     temp.alpha = static_cast<unsigned char>((right_boundary[i]-x)/std::abs(delta)*max_alpha);
                     if (valid_area.contains(p))
-                        assign_pixel(image(p.y(), p.x()),temp);
+                        assign_pixel(img[p.y()][p.x()],temp);
                 }
             }
             else // on the bottom side
             {
                 const long old_right_x = static_cast<long>(std::floor(right_boundary[i-1]));
                 for (long x = right_x+1; x <= old_right_x; ++x)
                 {
-                    const point p(x, i+top);
+                    const point p(x, i + valid_area.top());
                     rgb_alpha_pixel temp = alpha_pixel;
                     temp.alpha = static_cast<unsigned char>((right_boundary[i-1]-x)/delta*max_alpha);
                     if (valid_area.contains(p))
-                        assign_pixel(image(p.y(), p.x()),temp);
+                        assign_pixel(img[p.y()][p.x()],temp);
                 }
             }
         }
     }
 
-    template <typename image_type>
-    inline void draw_solid_convex_polygon (
+// ---------------------------------------------------------------------------------------
+
+    template <typename image_type, typename pixel_type>
+    void draw_solid_polygon (
         image_type& image,
-        const polygon& poly
-    ) { fill_convex_polygon(image, poly, 0); }
+        const polygon& poly,
+        const pixel_type& color,
+        const rectangle& area = rectangle(std::numeric_limits<long>::min(), std::numeric_limits<long>::min(),
+                                          std::numeric_limits<long>::max(), std::numeric_limits<long>::max())
+    )
+    {
+        image_view<image_type> img(image);
+        rectangle valid_area(get_rect(image).intersect(area));
+        const rectangle bounding_box = poly.get_rect();
+        if (bounding_box.intersect(valid_area).is_empty())
+            return;
+
+        valid_area = shrink_rect(valid_area.intersect(bounding_box), 1);
+
+        for (long y = valid_area.top(); y <= valid_area.bottom(); ++y)
+        {
+            // Compute the intersections with the scanline
+            std::vector<double> intersections;
```

Comment:
```suggestion
        std::vector<double> intersections;
        for (long y = valid_area.top(); y <= valid_area.bottom(); ++y)
        {
            // Compute the intersections with the scanline.
            intersections.clear();
```
Will be way faster since it avoids repeated reallocations.